### PR TITLE
JAMES-2287 Create a top level Blob storage API

### PR DESCRIPTION
**Pull request by aduprat**
Rebased & reworked version of #1234 
Every comments are solved, but:
- https://github.com/linagora/james-project/pull/1234#discussion_r161695796
- https://github.com/linagora/james-project/pull/1234#discussion_r161694674

### CODE REVIEW

File: server/blob/james-server-blob-api/pom.xml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mbaechler) [inline server/blob/james-server-blob-api/pom.xml]: maybe the directory name is a bit long, usually we would name it blob-api

File: server/blob/pom.xml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mbaechler) [inline server/blob/pom.xml]: usually, we don't create this intermediate level of pom (I don't know why). Maybe we should stick to what was done until now ?

File: server/blob/james-server-blob-cassandra/pom.xml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            <artifactId>junit-platform-launcher</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
```
Comment (mbaechler) [inline server/blob/james-server-blob-cassandra/pom.xml]: not in the right order

File: server/blob/james-server-blob-cassandra/pom.xml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            <artifactId>equalsverifier</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.platform</groupId>
+            <artifactId>junit-platform-launcher</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
```
Comment (mbaechler) [inline server/blob/james-server-blob-cassandra/pom.xml]: not in the right order

File: server/blob/james-server-blob-api/pom.xml [hunk 5]
```diff
@@ -33,6 +33,38 @@
...
+            <artifactId>equalsverifier</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.platform</groupId>
+            <artifactId>junit-platform-launcher</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
```
Comment (mbaechler) [inline server/blob/james-server-blob-api/pom.xml]: not in the right order

File: server/blob/james-server-blob-api/pom.xml [hunk 6]
```diff
@@ -33,6 +33,38 @@
...
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
```
Comment (mbaechler) [inline server/blob/james-server-blob-api/pom.xml]: same here

File: server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.blob.api;
+
+import java.util.concurrent.CompletableFuture;
+
+public interface ObjectStore {
+
+    CompletableFuture<BlobId> save(byte[] data);
```
Comment (mbaechler) [inline server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java]: Did you read my comment about using {Input,Output}stream instead ?

File: server/blob/james-server-blob-api/src/test/java/org/apache/james/blob/api/ObjectStoreContract.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+import org.junit.jupiter.api.Test;
+
+import com.google.common.base.Strings;
+
+public interface ObjectStoreContract {
```
Comment (mbaechler) [inline server/blob/james-server-blob-api/src/test/java/org/apache/james/blob/api/ObjectStoreContract.java]: should handle very big data

File: server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java (hunk 7 above)
Comment (chibenwa) [reply server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java]: IMO we could have two methods types: 
 - One dealing with byte[], one for streams.

Comment (mbaechler) [reply server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java]: what is the point, wrapping a byte[] in a stream is so easy

Comment (chibenwa) [reply server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java]: Yes, but you might  loose the "completable future part"

Comment (aduprat) [reply server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java]: Yes, I have read your comment.
But you don't have read what I have said:
* in the PR description
* in our internal chat

I took two different media to explain that, I'm a bit surprised that you are asking this question.

File: server/blob/pom.xml (hunk 2 above)
Comment (aduprat) [reply server/blob/pom.xml]: We are doing this when introducing new branches in the project tree.
Here are two examples:
- backend-common
- server/protocols/jmap-integration-testing/

File: server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java (hunk 7 above)
Comment (mbaechler) [reply server/blob/james-server-blob-api/src/main/java/org/apache/james/blob/api/ObjectStore.java]: The downside of using a new PR.

File: server/blob/pom.xml (hunk 2 above)
Comment (mbaechler) [reply server/blob/pom.xml]: ok but it's not the case in server/pom.xml it's why I told you

Comment (aduprat) [reply server/blob/pom.xml]: The difference is that server/pom.xml was not introduced recently.

**Comment by aduprat** [comment]
I push force the changes (due to blob project names changes), but I introduce fixup for other comments.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Investigating...

```
[a0ddaa75c94dcb5edeee3f0cdd39bae0a471fad4] Failed tests: 
[a0ddaa75c94dcb5edeee3f0cdd39bae0a471fad4]   ManageableMailQueueContract.clearShouldRemoveAllElements:473 
[a0ddaa75c94dcb5edeee3f0cdd39bae0a471fad4] Expecting empty but was:<[org.apache.james.queue.jms.JMSMailQueue$1$1@1c852c0f]>
```
test this please

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/DockerCassandraExtension.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+    private final DockerCassandraRule cassandraContainer;
+    private DockerCassandra dockerCassandra;
+
+    public DockerCassandraExtension() {
+        cassandraContainer = new DockerCassandraRule();
+    }
+
+    @Override
+    public void beforeEach(ExtensionContext context) throws Exception {
```
Comment (chibenwa) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/DockerCassandraExtension.java]: ... beforeAll + afterAll for performance sake.

File: mailbox/cassandra/pom.xml [hunk 10]
```diff
@@ -62,6 +62,16 @@
             <type>test-jar</type>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>blob-cassandra</artifactId>
```
Comment (chibenwa) [inline mailbox/cassandra/pom.xml]: That, should be with scope test. We do not want to enforce the cassandra blob store choice, do we?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/CassandraMailboxSessionMapperFactory.java [hunk 11]
```diff
@@ -23,6 +23,7 @@
 
 import org.apache.james.backends.cassandra.init.CassandraConfiguration;
 import org.apache.james.backends.cassandra.utils.CassandraUtils;
+import org.apache.james.blob.cassandra.CassandraBlobsDAO;
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/CassandraMailboxSessionMapperFactory.java]: That, should not appear in this class.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAOV2.java [hunk 12]
```diff
@@ -115,7 +116,7 @@ public static DAOAttachment from(Attachment attachment, BlobId blobId) {
     private static DAOAttachment fromRow(Row row) {
         return new DAOAttachment(
             AttachmentId.from(row.getString(ID)),
-            BlobId.from(row.getString(BLOB_ID)),
+            CassandraBlobId.from(row.getString(BLOB_ID)),
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAOV2.java]: That, should be a generic factory.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentMapper.java [hunk 13]
```diff
@@ -28,6 +28,7 @@
 
 import javax.inject.Inject;
 
+import org.apache.james.blob.cassandra.CassandraBlobsDAO;
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentMapper.java]: Should be an ObjectStore, not the DAO

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2Migration.java [hunk 14]
```diff
@@ -22,9 +22,9 @@
 import javax.inject.Inject;
 
 import org.apache.james.backends.cassandra.migration.Migration;
+import org.apache.james.blob.cassandra.CassandraBlobsDAO;
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2Migration.java]: Should be an object store, not the DAO

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobId.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.api;
+
+public interface BlobId {
+
```
Comment (chibenwa) [inline server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobId.java]: We SHOULD have a factory IMO.

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobsDAOTest.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Override
+    public ObjectStore testee() {
+        return testee;
+    }
+
+    @Override
+    public BlobId from(String blodIdAsString) {
+        return CassandraBlobId.from(blodIdAsString);
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobsDAOTest.java]: Rely on a factory instead.

**Comment by aduprat** [comment]
> Well... I see that my yesterday hard work did not receive attention, so I gonna do it again.

I wanted to do that, but in another review... Just merge the move, then enhance the code.
Actually the review is going to be too big, and is not finished, for example you forget to commit _CassandraObjectStore_.

**Comment by chibenwa** [comment]
```
[d3bd29f22fdf52f00f0808335985d603d7206b9f] Tests run: 65, Failures: 0, Errors: 1, Skipped: 6, Time elapsed: 51.329 sec <<< FAILURE! - in org.apache.james.queue.activemq.ActiveMQMailQueueTest
[d3bd29f22fdf52f00f0808335985d603d7206b9f] queueShouldPreserveName()  Time elapsed: 0.014 sec  <<< ERROR!
[d3bd29f22fdf52f00f0808335985d603d7206b9f] java.lang.RuntimeException: javax.jms.JMSException: Disposed due to prior exception
[d3bd29f22fdf52f00f0808335985d603d7206b9f] Caused by: javax.jms.JMSException: Disposed due to prior exception
[d3bd29f22fdf52f00f0808335985d603d7206b9f] Caused by: org.apache.activemq.transport.TransportDisposedIOException: Disposed due to prior exception
[d3bd29f22fdf52f00f0808335985d603d7206b9f] Caused by: org.apache.activemq.transport.TransportDisposedIOException: peer (vm://localhost#69) stopped.
```

test this please

**Comment by aduprat** [comment]
test this please
```
Tests in error: 
  PriorityMailQueueContract.tooBigPriorityShouldDefaultToMaximalPriority » Runtime
```

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**